Cleanup and add clarifications to DartPad pages (#3180)
dart-lang/site-www: src/tools/dartpad/index.md

description: The tool that lets you interactively play with Dart in a browser.
---

> DartPad is an [open source](https://github.com/dart-lang/dart-pad) tool

Author: I went with this, but I will give it a little bit more thought. Perhaps we could say something like the following, I'm not sure.

> DartPad is an open source tool, hosted [on GitHub](https://github.com/dart-lang/dart-pad),
> that lets you play with the Dart language in any modern browser.

lets you play with the Dart language in any modern browser.
DartPad is an [open source](https://github.com/dart-lang/dart-pad) tool
that lets you play with the Dart language in any modern browser.
Many pages in this site — especially [codelabs](/codelabs) —
have [embedded DartPads](#embedding).
To get a DartPad as big as your browser window, go to the

> <a href="{{site.dartpad}}" target="_blank" rel="noopener">DartPad

Reviewer: @kwalrath is the `rel="noopener"` required? IIUC, [noopener](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/noopener) is a security defense against untrusted content.

Author: It was a mistake for me to completely remove the `target="_blank"` as that allows the link to open in a new tab/window. I've added it back with Kramdown syntax to maintain that behavior.

According to the link you mentioned, this also "implicitly provides the same rel behavior as setting rel="noopener"" so we should be good.

  tips](/tools/dartpad/troubleshoot).
{{site.alert.end}}

Here's what DartPad looks like when configured to run Dart:

<img src="{% asset dartpad-hello.png @path %}" alt="Showcases what a Hello World application looks like in DartPad" />

## Library support

DartPad supports `dart:*` [core libraries](/guides/libraries) marked as

> multi-platform and web platform. When writing Flutter apps,
> DartPad also supports the `package:flutter` libraries.

Reviewer 2: nitpick:

```suggestion
multi-platform and web platform. When writing Flutter apps,
DartPad also supports the `package:flutter`
and `dart:ui` libraries.
```

Author: Good catch, thanks!

  tips](/tools/dartpad/troubleshoot).
{{site.alert.end}}

Here's what DartPad looks like when configured to run Dart:

> <img src="{% asset dartpad-hello.png @path %}" alt="Showcases what a Hello World application looks like in DartPad" />

Reviewer 3: ```suggestion
<img src="{% asset dartpad-hello.png @path %}" alt="Showcases what a Hello World app looks like in DartPad" />
```

## Library support

DartPad supports `dart:*` [core libraries](/guides/libraries) marked as
multi-platform and web platform. When writing Flutter apps,

> DartPad supports the `package:flutter`

Reviewer 3: ```suggestion
DartPad also supports the `package:flutter`
```

and `dart:ui` libraries.

DartPad doesn't support [deferred loading][] or any other libraries. 
For example, DartPad doesn't support using packages from 
the [pub.dev]({{site.pub}}) package repository. 

> (To find the status of package support, see [issue 901][].)

Reviewer 3: ```suggestion
(Package support might change; for details, see [issue 901][].)
```

  tips](/tools/dartpad/troubleshoot).
{{site.alert.end}}

> Here's what DartPad looks like when configured to run Dart:

Author: Giving it more thought, I agree this doesn't add any more clarity and isn't really necessary. Thanks!

multi-platform and web platform. It doesn't support those marked native
platform, and it doesn't support [deferred loading][].

If you are using Flutter, DartPad also supports the `package:flutter` libraries

> which are supported on the web platform.

Author: I looked into which libraries support web and as far as I could tell it actually seems all the top level provided `package:flutter` libraries seem to support web. Functionality which may not seems to be in other packages or plugins, but I don't know much about Flutter's setup there. 

I've updated the wording to no longer mention web support in relation to the Flutter libraries, but if someone has more information I'd be happy to include and link to it :)

> which are supported on the web platform.

Reviewer: Roping in @RedBrogdon and @johnpryan on this one. I worry about how the backend compiles stuff a lot more than how the front end presents the results of the compilation.

There are differing presentation of the editor panes for code that imports flutter versus code that doesn't. How we communicate that difference, I know not 😂

> It doesn't currently support using packages from the [pub.dev]({{site.pub}}) 

Author: Thanks for the suggestions! I've restructured this a bit, as when reading it all again I found that specifying that native libraries aren't supported is the reason we specify it supports "dart:* core libraries that are _marked as multi-platform and web platform_", so it didn't seem necessary to mention it in just a different way. Let me know if you think it's still worth a mention though.

   and confirm that you want to discard changes to the current pad.
   

2. Click the Dart logo, make sure that **HTML** support is disabled,
   and then click **Create**.
   

3. Change the code. For example, change the `main()` function
   to contain this code:  
   
   <!-- library-tour/string-tests/bin/main.dart -->
   {% prettify dart tag=pre+code %}
   for (var char in 'hello'.split('')) {
     print(char);
   }
   {% endprettify %}  
   
   As you type, DartPad shows hints, documentation,
   and autocomplete suggestions.
   

4. Click the **Format** button.  
   
   DartPad uses the [Dart formatter](/tools/dart-format)
   to ensure that your code has proper indentation, white space,
   and line wrapping.
   

5. Run your app.
   

6. If you didn't happen to have any bugs while you were entering the code,
   try introducing a bug.  

   For example, if you change `split` to `spit`,
   you get warnings at the bottom right of the window.
   If you run the app, a compilation error appears in the console.

## Checking Dart version info

The language features and APIs that DartPad supports depend on the
**Dart SDK** version that DartPad is currently using.

> You can find the active SDK's version at the bottom right of DartPad.

Reviewer 3: ```suggestion
You can find this SDK version at the bottom right of DartPad.
```

(I'm not sure that "active SDK" is clear enough, or that we need to introduce "active SDK" as a term.)

Author: Thanks for pointing this out, it definitely seems unnecessary.
